### PR TITLE
Add `--check` support for zipapps.

### LINKED ISSUE #2247: The Pex CLI should warn when it creates a PEX zip that requires zip64.
The python zipimporter used to execute a zipapp `__main__.py` in the zip cannot handle zips that require zip64 extensions (see: https://bugs.python.org/issue32959). As such, if the Pex CLI can detect when it creates such a zip, it should warn if the detection is approximate or probably error if it is definite; otherwise the error later at runtime is a mysterious:
```
/path/to/python: can't find '__main__' module in '/path/of/PEX'
```

At some point this will be fixed (presumably) for some CPython versions, in which case the logic gets trickier. The absolute worst thing to do here would be to warn or error for a PEX that works with a version of Python where this is no longer a limitation!

This does raise the question of whether or not this is an issue for PyPy - no clue.

### PR DESCRIPTION
**Pull request by jsirois**
By default Pex now warns for un-useable PEX zipapps it creates, but it
can be made to error or do nothing at all.

Fixes #2247

### CODE REVIEW

**Comment by jsirois** [comment]
These diskfulls were spurious, but now the new test creates a 4GB file and that appears to make diskfulls reliable. This PR will need #2255 to go in before it can go green it appears.

File: pex/pex_builder.py [hunk 1]
```diff
@@ -59,6 +61,64 @@ class Value(Enum.Value):
...
+                return None
+
+            try:
+                importer = zipimport.zipimporter(path)
+
+                # N.B.: The legacy `find_module` method returns the `zipimporter` instance itself on
+                # success and the `find_spec` method returns a `ModuleSpec` instance on success, but
+                # both return `None` on failure to find the module.
+                finder = "find_spec" if hasattr(importer, "find_spec") else "find_module"
+                if getattr(importer, finder)("__main__") is not None:
```
Comment (huonw) [inline pex/pex_builder.py:90]: Do you have a sense of the performance of this? I assume it won't be running any code within the `__main__` module, and thus will be some O(1) file manipulation/look-ups?

**Review by huonw — APPROVED** [review]
Makes sense to check in advance if possible 👍

File: pex/pex_builder.py (hunk 1 above)
Comment (jsirois) [reply pex/pex_builder.py:90]: Yeah, I actually didn't proceed with this PR until I had done that homework. It's ~100ms for the real case @peachanG hit in Pants (https://github.com/pantsbuild/pants/issues/19505) as documented here: https://github.com/pantsbuild/pex/issues/2247#issuecomment-1741506222

**Comment by jsirois** [comment]
> Makes sense to check in advance if possible.

Yeah, @peachanG's Pants issue was maybe the third to fifth time I recall someone running into this over the years. I was very worried about checking this robustly since you can build a PEX using an interpreter it will never actually run on. The IT added should catch when / if CPython adds support in 3.13 or later. At that point this currently robust solution has the potential to produce false positives and I'll need a solution.

File: pex/pex_builder.py (hunk 1 above)
Comment (jsirois) [reply pex/pex_builder.py:90]: Ok, yeah - worse than several seconds by a few orders of magnitude abd the check is also quicker than 100ms by a few order of magnitude (my vetting in the issue did not subtract Python start-up / import time of stdlib stuff which is ~all of the 100ms!). Re-running @peachanG's example at https://github.com/peachanG/pants-sample/tree/whisper with warm pantsd and warm pex cache (already ran `pants package ::` once):
```
$ pants package src/python/main/whisper:main --no-local-cache --pex-verbosity=1
...
pex:     Resolving requirements from lock file thirdparty/python/python-default.lock: 3334.0ms
pex:       Parsing requirements: 7.2ms
pex:       Resolving urls to fetch for 13 requirements from lock thirdparty/python/python-default.lock: 12.7ms
pex:       Downloading 41 distributions to satisfy 13 requirements: 27.5ms
pex:       Categorizing 41 downloaded artifacts: 0.3ms
pex:       Building 3 artifacts and installing 41: 3284.2ms
pex:         Building distributions for: ... 2.8ms
pex:         Calculating project names for direct requirements: ... 0.1ms
pex:         Installing 41 distributions: 3103.6ms
Saving PEX file to src.python.main.whisper/main.pex
pex: Zipping PEX file.
pex: Zipping PEX file.: 305577.9ms

09:10:46.86 [INFO] Wrote dist/src.python.main.whisper/main.pex
```

Now this case is the too large case. In the too many entries case you can imagine a smallish zip64 zip made up of >~65k files that zips relatively quickly. Adding some timing instrumentation to the IT added in this PR:
```
$ tox -qq -epy311 -- --devpi -sk 'test_check and not test_check_resolve_prerelease_transitive_dependencies_issue_1730' --quiet -lwarn
WARNING Logging configured at level WARN.
>>> Create too big file took 3.888437s
>>> Create too big zipapp took 5.139405s
>>> Check warn too-big.pyz.bust-cache took 0.000181s
>>> Check error too-big.pyz.bust-cache took 0.000035s
/home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py311/bin/python: can't find '__main__' module in '/tmp/pytest-of-jsirois/pytest-10/test_check0/too-big.pyz.bust-cache'
>>> Create too many entries zipapp took 0.775023s
>>> Check warn too-many-entries.pyz.bust-cache took 0.000094s
>>> Check error too-many-entries.pyz.bust-cache took 0.000029s
/home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py311/bin/python: can't find '__main__' module in '/tmp/pytest-of-jsirois/pytest-10/test_check0/too-many-entries.pyz.bust-cache'
>>> Create minimal zipapp took 0.000150s
>>> Check OK layout=zipapp check=none took 0.000002s
>>> Check OK layout=zipapp check=warn took 0.000088s
>>> Check OK layout=zipapp check=error took 0.000074s
>>> Check OK layout=packed check=none took 0.000001s
>>> Check OK layout=packed check=warn took 0.000001s
>>> Check OK layout=packed check=error took 0.000001s
>>> Check OK layout=loose check=none took 0.000001s
>>> Check OK layout=loose check=warn took 0.000001s
>>> Check OK layout=loose check=error took 0.000001s
.
1 passed, 685 deselected in 11.40s
```

These represent worse cases and you get:
+ single too big file + __main__.py with no Pex in the way building: ~5s build, <200us check
+ ~65k small python files + __main__.py with no Pex in the way building: ~1s build, <100us to check

**Comment by jsirois** [comment]
Alright, wrapping this up.

With a `tox -epackage -- --serve` and this edit to @peachanG's sample repo:
```diff
$ git diff pants.toml
diff --git a/pants.toml b/pants.toml
index 298039b..369b729 100644
--- a/pants.toml
+++ b/pants.toml
@@ -1,3 +1,8 @@
+[DEFAULT]
+pex_version = "v2.1.147+5-ge511cf6"
+pex_fp = "5035cbb22f5d2daf3db7677e04fecd4dc32a6a50167d2caa4ef0cf44c03cfa86"
+pex_size = 4203379
+
 [GLOBAL]
 pants_version = "2.17.0"
 backend_packages.add = [
@@ -9,6 +14,16 @@ backend_packages.add = [
 interpreter_constraints = ["CPython==3.8.*"]
 enable_resolves = true

+[pex-cli]
+url_template = "http://127.0.0.1:41307/pex"
+version = "%(pex_version)s"
+known_versions = [
+  "%(pex_version)s|macos_arm64|%(pex_fp)s|%(pex_size)s",
+  "%(pex_version)s|macos_x86_64|%(pex_fp)s|%(pex_size)s",
+  "%(pex_version)s|linux_x86_64|%(pex_fp)s|%(pex_size)s",
+  "%(pex_version)s|linux_arm64|%(pex_fp)s|%(pex_size)s"
+]
+
 [source]
 root_patterns = ["src/python"]
```

You get:
```
$ pants package src/python/main/whisper:main --pex-verbosity=1
...
Saving PEX file to src.python.main.whisper/main.pex
pex: Zipping PEX file.
pex: Zipping PEX file.: 297668.8ms
/home/jsirois/.cache/pants/named_caches/pex_root/installed_wheels/153189a206d2817495e79d53afe6413953c00e2e04bc7a810f214fa1860de6d3/pex-2.1.147-py2.py3-none-any.whl/pex/pex_builder.py:114: PEXWarning: The PEX zip at src.python.main.whisper/main.pex~ is not a valid zipapp: Could not find the `__main__` module.
This is likely due to the zip requiring ZIP64 extensions due to size or the
number of file entries or both. You can work around this limitation in Python's
`zipimport` module by re-building the PEX with `--layout packed` or
`--layout loose`.
  pex_warnings.warn(message)

10:22:53.78 [INFO] Wrote dist/src.python.main.whisper/main.pex
```

Unfortunately that is both noisy (lots of Pex verbose output snipped by the ...) and non-default. By default the warning is suppressed by Pants:
```
$ pants package src/python/main/whisper:main
10:23:54.46 [INFO] Completed: Building local_dists.pex
10:29:09.20 [INFO] Completed: Building 13 requirements for src.python.main.whisper/main.pex from the thirdparty/python/python-default.lock resolve: nvidia-cublas-cu11==11.10.3.66; platform_system == "Linux" and platform_machine ==... (970 characters truncated)
10:29:10.16 [INFO] Wrote dist/src.python.main.whisper/main.pex
```

**Pull request merged**